### PR TITLE
Fix generate field SQL for MySQL virtual columns

### PR DESCRIPTION
**Pull request by mvorisek**
Fix issue #13854, based on official syntax:
column_definition: https://dev.mysql.com/doc/refman/5.7/en/create-table.html

### CODE REVIEW

**Comment by williamdes** [comment]
Note for myself: add percona db to if statement if testing is okay using percona

@mvorisek can you please add percona db?, I will do it if you did not have time

**Comment by mvorisek** [comment]
@williamdes Please do it, I do not have a Percona DB to test.

File: libraries/classes/Table.php [hunk 1]
```diff
@@ -534,17 +534,25 @@ public static function generateFieldSpec(
             }
         }
 
-        if ($virtuality) {
-            $query .= ' AS (' . $expression . ') ' . $virtuality;
-        } else {
+        // if column is virtual, check if server type is Mysql as only Mysql server
+        // supports extra column properties
+        $virtualityMysql = $virtuality && Util::getServerType() === 'MySQL';
```
Comment (williamdes) [inline libraries/classes/Table.php]:
```suggestion
        $virtualityMysql = $virtuality && in_array(Util::getServerType(), ['MySQL', 'Percona Server']);
```

**Review by williamdes — APPROVED** [review]

File: libraries/classes/Table.php (hunk 1 above)
Comment (williamdes) [reply libraries/classes/Table.php]: @mvorisek Can you click on "commit suggestion" ?

**Comment by williamdes** [comment]
@sudo-bot Deploy pull-request

**Comment by mvorisek** [comment]
@williamdes Can you please redeploy and test with both server types?

**Comment by williamdes** [comment]
@sudo-bot Deploy PR

Done @mvorisek :)

**Comment by mvorisek** [comment]
@sudo-bot Deploy PR

**Comment by williamdes** [comment]
@sudo-bot /deploy

**Comment by williamdes** [comment]
@mvorisek Can you please rebase the PR onto latest `QA_4_8` ?

File: libraries/classes/Table.php [hunk 2]
```diff
@@ -534,17 +534,29 @@ public static function generateFieldSpec(
...
+        // supports no extra virtual column properties except CHARACTER SET for text column types
+        $isVirtualColMariaDB = $virtuality && Util::getServerType() === 'MariaDB';
+
+        $matches = preg_match(
+            '@^(TINYTEXT|TEXT|MEDIUMTEXT|LONGTEXT|VARCHAR|CHAR|ENUM|SET)$@i',
+            $type
+        );
+        if (! empty($collation) && $collation != 'NULL' && $matches) {
+            $query .= Util::getCharsetQueryPart(
+                $isVirtualColMariaDB ? preg_replace('~_.+~s', '', $collation) : $collation,
```
Comment (williamdes) [inline libraries/classes/Table.php]: Can you add a comment on what id does ?

**Review by williamdes — APPROVED** [review]
Perfect !

Seems PerconaDB still has the BLOB issue, maybe there is something missing to handle like MySQL

File: libraries/classes/Table.php (hunk 2 above)
Comment (mvorisek) [reply libraries/classes/Table.php]: I found out that `CHARACTER SET` attr. is accepted by Maria DB, so it strips the `COLLACTION`. Documented in the code: https://github.com/phpmyadmin/phpmyadmin/blob/98814b1a250e877f4b782ae7f2ea9e52d8530932/libraries/classes/Table.php#L541

**Comment by mvorisek** [comment]
> Seems PerconaDB still has the BLOB issue
Can you explain? The original BLOB issue https://github.com/phpmyadmin/phpmyadmin/issues/13854#issuecomment-493714165 seems to be resolved.

**Comment by williamdes** [comment]
> The original BLOB issue #13854 (comment) seems to be resolved.

Yes, but not on PerconaDB (see deployed server (Percona 1 & 2))

File: libraries/classes/Table.php (hunk 2 above)
Comment (williamdes) [reply libraries/classes/Table.php]: Thanks !

**Comment by mvorisek** [comment]
PerconaDB behaves like MySQL so the code works. But there is some issue with the GUI as no virtuallity in the GUI is offered at all for PerconaDB. Any idea where the GUI condition is located and should the fix be part of this PR?

> Can`you please rebase the PR onto latest QA_4_8 ?

I can, but how? The PR is opened for `mvorisek:master` -> ` phpmyadmin:master`.

**Comment by williamdes** [comment]
> I can, but how? The PR is opened for mvorisek:master -> phpmyadmin:master.

Sorry, I forgot about that, again..
Can you please create a branch from QA and cherry-pick your changes (do not forget to change the array syntax (`[ ]` to `array()` ) and open another PR

I prefer to ask you because if not, I will have to cherry-pick your changes and edit them when merging ..

let's forget about the PerconaDB issue and merge your work into QA :)

**Comment by williamdes** [comment]
@mvorisek The PR works everywhere now, see 966902b7a546adcb71b8eb88ea9c096ede299bde

**Comment by williamdes** [comment]
Merged into master with commit de4bc0751b858464d2be14090478e078ef3dc35f

**Pull request closed without merging**